src/data/nat/factorization.lean

```diff
@@ -374,6 +374,82 @@ begin
   exact support_add_eq (factorization_disjoint_of_coprime hab),
 end
 
+/-! ### Factorization and `is_square` -/
+
+lemma is_square_iff_factorization_even {m : ℕ} :
+  is_square m ↔ ∀ (p : ℕ), even (m.factorization p) :=
+begin
+  rcases eq_or_ne m 0 with rfl | hm_zero,
+  { simp, },
+  { split,
+    { rintros ⟨c, hc⟩,
+      by_cases hc' : c = 0,
+      { rw [hc', mul_zero] at hc,
+        rw hc,
+        simp only [factorization_zero, finsupp.coe_zero, pi.zero_apply, forall_const, even_zero], },
+      { intros p,
+        rw [hc, nat.factorization_mul hc' hc', finsupp.coe_add],
+        exact even_add_self _, }, },
+    { intros hp,
+      use m.factorization.prod (λ p i, p ^ (i / 2)),
+      rw ← finsupp.prod_mul,
+      conv { to_lhs, rw ← nat.factorization_prod_pow_eq_self hm_zero, },
+      apply finsupp.prod_congr,
+      intros x hx,
+      rcases hp x with ⟨c, hc⟩,
+      rw [hc, ←two_mul],
+      simp only [mul_div_right, succ_pos'],
+      rw [←pow_add, ←two_mul], }, },
+end
+
+lemma is_square.of_mul_left {m n : ℕ} (hmn : m.coprime n) :
```

Comment:
This might be mistaken for `is_square (m * n) → is_square n → is_square m`. Can you come up with a better name? What about `is_square.of_coprime_left`?

Reply:
Oh hmmm. It was meant to be in the same vein as the name `squarefree.of_mul_left` which has a similar coprimality condition.

However, perhaps `is_square.of_coprime_of_mul_left`?

Reply:
I was happy with the shorter name, but it's up to you!

Reply:
I feel like `mul` is important for searching, so let's keep the longer name!